src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+            H5FD_mem_t              *io_types_ptr  = NULL;
+            uint32_t                 final_vec_len = iovec_len;
+            haddr_t                 *io_addrs_ptr  = NULL;
+            size_t                  *io_sizes_ptr  = NULL;
+
+            /* Setup index into I/O vectors for this I/O operation */
+            io_types_ptr = &io_types[ioreq_idx * iovec_len];
+            io_addrs_ptr = &io_addrs[ioreq_idx * iovec_len];
+            io_sizes_ptr = &io_sizes[ioreq_idx * iovec_len];
+            io_bufs_ptr  = &io_bufs[ioreq_idx * iovec_len];
+
+            /* Skip 0-sized I/Os */
+            for (size_t vec_idx = 0; vec_idx < iovec_len; vec_idx++)
+                if (io_sizes_ptr[vec_idx] == 0)
+                    final_vec_len--;
+
+            if (io_type == IO_TYPE_WRITE) {
+                /* Make vector write call to VFD controlling subfiles */
+                if (H5FD_write_vector(file_ptr->sf_file, final_vec_len, io_types_ptr, io_addrs_ptr,
+                                      io_sizes_ptr, (const void **)io_bufs_ptr) < 0)
+                    H5_SUBFILING_GOTO_ERROR(H5E_IO, H5E_WRITEERROR, FAIL, "write to subfile failed");
+
+                /*
+                 * Mirror superblock writes to the stub file so that
+                 * legacy HDF5 applications can check what type of
+                 * file they are reading
+                 */
+                if (H5FD__subfiling_mirror_writes_to_stub(file_ptr, final_vec_len, io_types_ptr, io_addrs_ptr,
+                                                          io_sizes_ptr, (const void **)io_bufs_ptr) < 0)
+                    H5_SUBFILING_GOTO_ERROR(H5E_IO, H5E_WRITEERROR, FAIL,
+                                            "mirrored write to stub file failed");
+            }
+            else {
+                if (!rank0_bcast || (file_ptr->mpi_rank == 0)) {
+                    /* Make vector read call to VFD controlling subfiles */
+                    if (H5FD_read_vector(file_ptr->sf_file, final_vec_len, io_types_ptr, io_addrs_ptr,
+                                         io_sizes_ptr, (void **)io_bufs_ptr) < 0)
+                        H5_SUBFILING_GOTO_ERROR(H5E_IO, H5E_READERROR, FAIL, "read from subfile failed");
+                }
+            }
+        }
+
+        if (rank0_bcast && (file_ptr->mpi_size > 1)) {
+            size_t size;
+
+            assert(io_type == IO_TYPE_READ);
+
+            extend_sizes = false;
+            for (size_t i = 0; i < io_count; i++) {
+                if (!extend_sizes) {
+                    if ((i > 0) && (sizes[i] == 0)) {
+                        extend_sizes = true;
+                        size         = sizes[i - 1];
+                    }
+                    else {
+                        size = sizes[i];
+                    }
+                }
+
+                H5_CHECK_OVERFLOW(size, size_t, int);
+                if (MPI_SUCCESS != MPI_Bcast(bufs[i].vp, (int)size, MPI_BYTE, 0, file_ptr->comm))
```

Comment:
Maybe not the focus of this PR but how hard would it be to accumulate these into a single bcast?

Reply:
Indeed, this was mostly just copied over from the previous code, but it could probably be handled better with a derived type instead of multiple bcasts.